### PR TITLE
implemented rooms API

### PR DESCRIPTION
**Pull request by danbaranov**
1. Rewrote server.js to Typescript.
2. Implemented simple Rooms API

### CODE REVIEW

File: src/api/rooms.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    this.prisma = new PrismaClient()
+    this.router.use(express.json())
+    this.intializeRoutes()
+  }
+
+  public intializeRoutes() {
+    this.router.get(this.path, this.getAllRooms)
+    this.router.post(
+      this.path,
+      [body('name').notEmpty().trim(), body('definition').isJSON()],
```
Comment (zeitgeist) [inline src/api/rooms.ts:21]: 👍 Shall we proceed discussing the design of a proper benchmarking type definition through a follow up PR?

File: prisma/schema.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE rooms(
+    id SERIAL PRIMARY KEY NOT NULL,
```
Comment (zeitgeist) [inline prisma/schema.sql:2]: Can we change this to a uuid so, at least, a room ID cannot easily be guessed / scraped easily?

**Review by zeitgeist — APPROVED** [review]
Thank you, Denis!

**Comment by danbaranov** [comment]
I've replaced serial with uuid

File: src/api/rooms.ts (hunk 1 above)
Comment (danbaranov) [reply src/api/rooms.ts:21]: Yes, I think we need to make a sync call regarding this

**Comment by felixniemeyer** [comment]
Looks tidy!

**Comment by felixniemeyer** [comment]
From a technical point of view I'm very happy with this merge request. 

Here is my understanding of the whole process from creating rooms till executing mpc computations: 
I imagine there will be some kind of ID for rooms. 
If you create a room you choose or receive the ID for that room. 
You invite others by giving them the ID and maybe a participation token. 
Now everybody who receives the ID (+ participation token) will be able to join the room.
To join, participants will first connect to the rooms api server, which returns the session config. 
Using this config he will connect to and communicate with a MPC coordination server (jiff).
Once everyone is connected simultaneously, the mpc computation will be executed.  
  
Now, from an idealistic point of view I'm concerned about implementing the distribution of session configs by means of a central server while it would also be possible to simply share the whole session config instead of just an ID. 

What's your thoughts on that?

**Comment by danbaranov** [comment]
In this case this API is completely redundant. But what I'd love to see is how the benchmarking actually being used. I know contradicts a bit with the idea of ultimate privacy but as soon as it'a a demo/POC I don't think it's anything bad.

Another thing, and more important one, is if we go with more sophisticated ways of preventing of cheating when I can participate in a majority of secessions to learn about counter-party inputs. In this case we can implement invitations with secret participation tokens or something like that. In that case it's quite tricky, if even possible, to implement this without any server side API.

So the question is more about how far we'd like to go with it.

**Comment by felixniemeyer** [comment]
:+1:

**Comment by zeitgeist** [comment]
> On 30. Oct 2020, at 11:38, fairlix <notifications@github.com> wrote:
> 
> 
> Now, from an idealistic point of view I'm concerned about implementing the distribution of session configs by means of a central server while it would also be possible to simply share the whole session config instead of just an ID.
> 

> 
> What's your thoughts on that?
> 
> 

1. It is a demo utility at this point and I do not regard this as a problem whatsoever
1. We can always switch from an UUID as a string identifier to an “onomatopoetic” ID which is the session description encoded
1. I think it’s great to have a means to simply join an existing session

**Pull request merged**